rust/lance-encoding/src/buffer.rs

```diff
@@ -205,10 +221,20 @@ impl LanceBuffer {
 
     /// Reinterprets a LanceBuffer into a Vec<T>
     ///
-    /// Unfortunately, there is no way to do this safely in Rust without a copy, even if
-    /// the source is Vec<u8>.
+    /// If the underlying buffer is not properly aligned, this will involve a copy of the data
     pub fn borrow_to_typed_slice<T: ArrowNativeType>(&mut self) -> impl AsRef<[T]> {
-        ScalarBuffer::<T>::from(self.borrow_and_clone().into_buffer())
+        let align = std::mem::align_of::<T>();
+        let is_aligned = self.as_ptr().align_offset(align) == 0;
+
+        if is_aligned {
```

Comment:
I think we may want to introduce a panic here when `lance_buffer`'s alignment is not equal to the alignment needed for the output type.
one reason is that we are implicitly dropping data here:
```rust
            let num_values = self.len() / std::mem::size_of::<T>();
            let vec = Vec::<T>::with_capacity(num_values);
            let mut bytes = MutableBuffer::from(vec);
            bytes.extend_from_slice(self);
```
for example, I can do:
```rust
    #[test]
    fn test_borrow_to_typed_slice() {
        let mut lance_buf = LanceBuffer::from([1, 0, 1, 1, 2, 1, 3].to_vec());
        let a = lance_buf.borrow_to_typed_slice::<u16>();
        println!("a: {:?}", a.as_ref());

    }
```
and see the value `3` in `[1, 0, 1, 1, 2, 1, 3]` is dropped(because `(a / b) * b != a` when do integer divide.

another reason is that I think we only do `borrow_to_typed_slice` when we know the underlying buffer if originated from the type we want

Reply:
@broccoliSpicy ah, good point.  I'll add a check for that.

Reply:
Ok, I added a check and panic if we aren't evenly divisible.  I've added some test cases to verify both this behavior and also to verify we don't copy if we are aligned and we do copy if we aren't aligned.